### PR TITLE
Fix Update-with-Start for closed workflow

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

Fixed bug where UwS does not work for a closed workflow even though the reuse policy is `ALLOW_DUPLICATE`.

## Why?

Bug.

## How did you test it?

Added new tests.

## Potential risks

I kept the changes to the non-UwS path to an _absolute_ minimum. I don't believe it changes the non-UwS behavior in any way.

### CODE REVIEW

File: service/history/api/multioperation/api.go [hunk 1]
```diff
@@ -162,7 +175,7 @@ func Invoke(
 	}
 
 	// workflow was already started, ...
-	if currentWorkflowLease != nil {
+	if currentWorkflowLease != nil && currentWorkflowLease.GetMutableState().IsWorkflowExecutionRunning() {
```
Comment (stephanos) [inline service/history/api/multioperation/api.go:178]: Turns out this check is important ...

File: service/history/api/create_workflow_util.go [hunk 2]
```diff
@@ -59,7 +59,7 @@ type (
 		RunID            string
 		LastWriteVersion int64
 	}
-	CreateLeaseFunc func(shard.Context, workflow.MutableState) (WorkflowLease, error)
+	CreateLeaseFunc func(WorkflowLease, shard.Context, workflow.MutableState) (WorkflowLease, error)
```
Comment (stephanos) [inline service/history/api/create_workflow_util.go:62]: As a quick fix, I had to add another parameter here for the case for the lease actually already exists. This is not pretty, but it works.

File: service/history/api/startworkflow/api.go [hunk 3]
```diff
@@ -424,7 +427,7 @@ func (s *Starter) resolveDuplicateWorkflowID(
 	case err != nil:
 		return nil, StartErr, err
 	case currentExecutionUpdateAction == nil:
-		return nil, StartErr, nil
+		return nil, StartNew, nil
```
Comment (stephanos) [inline service/history/api/startworkflow/api.go:430]: This was incorrect, it is actually starting a new workflow.

File: service/history/api/startworkflow/api.go [hunk 4]
```diff
@@ -491,7 +494,7 @@ func (s *Starter) resolveDuplicateWorkflowID(
 	case consts.ErrWorkflowCompleted:
 		// current workflow already closed
 		// fallthough to the logic for only creating the new workflow below
-		return nil, StartErr, nil
+		return nil, StartNew, nil
```
Comment (stephanos) [inline service/history/api/startworkflow/api.go:497]: This was incorrect, it is actually starting a new workflow.

File: service/history/api/startworkflow/api.go [hunk 5]
```diff
@@ -348,6 +347,10 @@ func (s *Starter) createAsCurrent(
 	creationParams *creationParams,
 	currentWorkflowConditionFailed *persistence.CurrentWorkflowConditionFailedError,
 ) error {
+	// TODO(stephanos): remove this hack - this is here for Update-with-start to reapply the Update
+	if _, err := s.createLeaseFn(creationParams.workflowLease, s.shardContext, nil); err != nil {
+		return err
+	}
```
Comment (stephanos) [inline service/history/api/startworkflow/api.go:353]: Note that his will now call `NewWorkflowLeaseAndContext` twice for a regular workflow start. But that should be fine, as it has no side effects.

File: service/history/api/create_workflow_util.go [hunk 6]
```diff
@@ -170,9 +170,14 @@ func NewWorkflowWithSignal(
...
 // NOTE: must implement CreateLeaseFunc.
 func NewWorkflowLeaseAndContext(
+	existingLease WorkflowLease,
 	shardCtx shard.Context,
 	ms workflow.MutableState,
 ) (WorkflowLease, error) {
+	// TODO(stephanos): remove this hack
+	if existingLease != nil {
+		return existingLease, nil
+	}
```
Comment (stephanos) [inline service/history/api/create_workflow_util.go:180]: This makes sure this is a noop when called with an existing lease. Which only happens in exactly one case (see below).

File: service/history/api/startworkflow/api.go [hunk 7]
```diff
@@ -348,6 +347,11 @@ func (s *Starter) createAsCurrent(
 	creationParams *creationParams,
 	currentWorkflowConditionFailed *persistence.CurrentWorkflowConditionFailedError,
 ) error {
+	// TODO(stephanos): remove this hack
+	// This is here for Update-with-Start to reapply the Update after it was aborted previously.
+	if _, err := s.createOrUpdateLeaseFn(creationParams.workflowLease, s.shardContext, nil); err != nil {
+		return err
+	}
```
Comment (stephanos) [inline service/history/api/startworkflow/api.go:354]: Note that his will now call NewWorkflowLeaseAndContext twice for a regular workflow start. But that is fine, as I added a special case to make that a noop there.

Comment (yycptt) [reply service/history/api/startworkflow/api.go:354]: do you mean `creationParams.workflowLease` is always nil?

**Review by yycptt — APPROVED** [review]
Approving as it doesn't change existing start workflow execution logic.

**Comment by stephanos** [comment]
I'll merge so it can be tested on the nightlies; and will think about whether there's a better way to do it.

**Pull request merged**